File: pages/homepage.py
from utils.drag_drop_mixin import DragDropMixin
from pages.base_page import BasePage

from PySide6.QtCore import QEasingCurve, QObject, QPropertyAnimation, Qt, Signal
from PySide6.QtWidgets import QSizePolicy, QLabel

from widgets.svg_button import SvgButton
from core.page_router import page_router

from resources.svgs import settings_icon, app_center_icon
from core.colors import WHITE, COLOR_TRANSPARENT, NEUTRAL_4, to_qss_color

# 拖拽提示样式（颜色统一由 colors.py 管理）
_DROP_HINT_IDLE_QSS = (
    f"color: {to_qss_color(COLOR_TRANSPARENT)}; font-size: 18px; "
    f"border: 2px dashed {to_qss_color(NEUTRAL_4)}; border-radius: 5px; "
)
_DROP_HINT_ACTIVE_QSS = (
    f"color: {to_qss_color(WHITE)}; font-size: 18px; "
    f"border: 2px dashed {to_qss_color(NEUTRAL_4)}; border-radius: 5px; "
)

class OnDragEvent(QObject):
    on_drag_event = Signal(bool)

on_drag_bus = OnDragEvent()

class HomePage(DragDropMixin, BasePage):
    PAGE_NAME = "home"
    TITLE = "Home"

    def __init__(self, parent=None):
        super().__init__(parent)
        self.init_drag_drop()

        self.target_size = (140, 50)

        layout = self.set_main_layout('h')
        assert layout is not None

        # 设置按钮 (齿轮)
        setting_btn = SvgButton(size=36, icon_size=22, svg_data=settings_icon, enable_rotation=True)
        setting_btn.clicked.connect(lambda: page_router.immediate_switch("setting"))

        app_center_btn = SvgButton(size=36, icon_size=22, svg_data=app_center_icon, enable_rotation=True)
        app_center_btn.clicked.connect(lambda: page_router.immediate_switch("module_center"))

        self.drop_hint_label = QLabel("Drag here", self)
        self.drop_hint_label.setAlignment(Qt.AlignmentFlag.AlignCenter)
        self.drop_hint_label.setStyleSheet(_DROP_HINT_IDLE_QSS)

        self.drop_hint_label.setSizePolicy(QSizePolicy.Policy.Expanding, QSizePolicy.Policy.Expanding)
        self.drop_hint_label.setMaximumWidth(0) 

        layout.addWidget(setting_btn)
        layout.addWidget(app_center_btn)
        layout.addWidget(self.drop_hint_label)

        self.drop_anim = QPropertyAnimation(self.drop_hint_label, b"maximumWidth")
        self.drop_anim.setDuration(800)
        self.drop_anim.setEasingCurve(QEasingCurve.Type.OutQuart)

    def on_drag_enter(self):
        self.drop_hint_label.setStyleSheet(_DROP_HINT_ACTIVE_QSS)

        on_drag_bus.on_drag_event.emit(True)
        self.drop_anim.stop()
        self.drop_anim.setEndValue(0)
        self.drop_anim.start()

    def on_drag_leave(self):
        on_drag_bus.on_drag_event.emit(False)
        self.drop_anim.stop()
        self.drop_anim.setEndValue(0)
        self.drop_anim.start()
        
    # def on_files_dropped(self, file_paths: list[str]):
    #     pass

File: pages/homepage.cpp
#include "pages/homepage.h"

#include "core/page_router.h"
#include "resources/colors.h"
#include "resources/svgs.h"
#include "widgets/svg_button.h"

#include <QDragEnterEvent>
#include <QDragLeaveEvent>
#include <QDropEvent>
#include <QBoxLayout>
#include <QEasingCurve>
#include <QMimeData>
#include <QSizePolicy>
#include <QUrl>

// ==================== OnDragBus ====================

OnDragBus& OnDragBus::instance()
{
    static OnDragBus bus;
    return bus;
}

// ==================== HomePage ====================

namespace {
// 拖拽提示样式（颜色统一由 colors 管理）
QString dropHintIdleQss()
{
    return QStringLiteral("color: %1; font-size: 18px; "
                           "border: 2px dashed %2; border-radius: 5px; ")
        .arg(Colors::toQssColor(Colors::transparent()), Colors::toQssColor(Colors::neutral4()));
}

QString dropHintActiveQss()
{
    return QStringLiteral("color: %1; font-size: 18px; "
                           "border: 2px dashed %2; border-radius: 5px; ")
        .arg(Colors::toQssColor(Colors::white()), Colors::toQssColor(Colors::neutral4()));
}
} // namespace

HomePage::HomePage(QWidget* parent)
    : BasePage(parent)
{
    setAcceptDrops(true);

    setTargetSize(QSize(140, 50));

    QBoxLayout* layout = setMainLayout('h');

    // 设置按钮 (齿轮)
    auto* settingBtn = new SvgButton(nullptr, 36, 22, Svgs::settingsIcon(), QColor(), true);
    connect(settingBtn, &SvgButton::clicked, this,
            []() { PageRouter::instance().immediateSwitch(QStringLiteral("setting")); });

    auto* appCenterBtn = new SvgButton(nullptr, 36, 22, Svgs::appCenterIcon(), QColor(), true);
    connect(appCenterBtn, &SvgButton::clicked, this,
            []() { PageRouter::instance().immediateSwitch(QStringLiteral("module_center")); });

    m_dropHintLabel = new QLabel(QStringLiteral("Drag here"), this);
    m_dropHintLabel->setAlignment(Qt::AlignmentFlag::AlignCenter);
    m_dropHintLabel->setStyleSheet(dropHintIdleQss());
    m_dropHintLabel->setSizePolicy(QSizePolicy::Policy::Expanding, QSizePolicy::Policy::Expanding);
    m_dropHintLabel->setMaximumWidth(0);

    layout->addWidget(settingBtn);
    layout->addWidget(appCenterBtn);
    layout->addWidget(m_dropHintLabel);

    m_dropAnim = new QPropertyAnimation(m_dropHintLabel, QByteArrayLiteral("maximumWidth"), this);
    m_dropAnim->setDuration(800);
    m_dropAnim->setEasingCurve(QEasingCurve::OutQuart);
}

// ---- 拖放事件（对应 DragDropMixin）----

void HomePage::dragEnterEvent(QDragEnterEvent* event)
{
    // 同时允许 带有文件链接(Urls) 或 纯文本(Text) 的内容拖入
    if (event->mimeData()->hasUrls() || event->mimeData()->hasText()) {
        event->acceptProposedAction();
        onDragEnter();
    } else {
        event->ignore();
    }
}

void HomePage::dragLeaveEvent(QDragLeaveEvent* event)
{
    onDragLeave();
    event->accept();
}

void HomePage::dropEvent(QDropEvent* event)
{
    onDragLeave();   // 放下通常也意味着离开拖拽状态

    const QMimeData* mimeData = event->mimeData();
    if (mimeData->hasUrls()) {
        QStringList filePaths;
        const QList<QUrl> urls = mimeData->urls();
        for (const QUrl& url : urls)
            filePaths.append(url.toLocalFile());
        onFilesDropped(filePaths);
        event->acceptProposedAction();
    } else if (mimeData->hasText()) {
        onTextDropped(mimeData->text());
        event->acceptProposedAction();
    }
}

void HomePage::onDragEnter()
{
    m_dropHintLabel->setStyleSheet(dropHintActiveQss());
    emit OnDragBus::instance().onDragEvent(true);
    m_dropAnim->stop();
    m_dropAnim->setEndValue(0);
    m_dropAnim->start();
}

void HomePage::onDragLeave()
{
    emit OnDragBus::instance().onDragEvent(false);
    m_dropAnim->stop();
    m_dropAnim->setEndValue(0);
    m_dropAnim->start();
}
